app/src/androidTest/java/org/mozilla/focus/activity/PullDownToRefreshTest.java

```diff
@@ -0,0 +1,129 @@
+import static android.support.test.espresso.matcher.ViewMatchers.withClassName;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withParent;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+import static android.support.test.espresso.web.assertion.WebViewAssertions.webMatches;
+import static android.support.test.espresso.web.sugar.Web.onWebView;
+import static android.support.test.espresso.web.webdriver.DriverAtoms.findElement;
+import static android.support.test.espresso.web.webdriver.DriverAtoms.getText;
+import static android.support.test.espresso.web.webdriver.DriverAtoms.webClick;
+import static junit.framework.Assert.assertTrue;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.endsWith;
+import static org.mozilla.focus.activity.PullDownToRefreshTest.withCustomConstraints;
+import static org.mozilla.focus.activity.TestHelper.progressBar;
+import static org.mozilla.focus.activity.TestHelper.waitingTime;
+import static org.mozilla.focus.activity.TestHelper.webPageLoadwaitingTime;
+import static org.mozilla.focus.activity.helpers.EspressoHelper.navigateToWebsite;
+import static org.mozilla.focus.activity.helpers.EspressoHelper.openSettings;
+
+@RunWith(AndroidJUnit4.class)
+public class PullDownToRefreshTest {
+    private static final String MOZILLA_WEBSITE_SLOGAN_SELECTOR = ".content h2";
+    private static final String MOZILLA_WEBSITE_SLOGAN_TEXT = "Internet for people,\nnot profit.";
+
+    @Rule
+    public ActivityTestRule<MainActivity> mActivityTestRule = new MainActivityFirstrunTestRule(false);
+
+    @ClassRule
+    public static final LocaleTestRule localeTestRule = new LocaleTestRule();
+
+    @Test
+    public void pullDownToRefreshTest() throws InterruptedException, UiObjectNotFoundException, IOException {
+        // Go to a website
+        navigateToWebsite(("www.mozilla.org"));
+
+        TestHelper.progressBar.waitForExists(webPageLoadwaitingTime);
+        Assert.assertTrue(TestHelper.progressBar.exists());
+        //Assert.assertTrue(TestHelper.progressBar.waitUntilGone(webPageLoadwaitingTime));
+
+        // Check that the webview is loaded
+        assertWebsiteUrlContains("mozilla.org");
+
+        onWebView()
+                .withElement(findElement(Locator.CSS_SELECTOR, MOZILLA_WEBSITE_SLOGAN_SELECTOR))
+                .check(webMatches(getText(), containsString(MOZILLA_WEBSITE_SLOGAN_TEXT)));
+
+        // Swipe down to refresh, spinner is shown (2nd child) and progress bar is shown
+        onView(withId(R.id.swipe_refresh))
+                .perform(withCustomConstraints(swipeDown(), isDisplayingAtLeast(85)))
+                .check(matches(hasChildCount(2)));
+
+        //Assert.assertTrue(TestHelper.progressBar.waitForExists(webPageLoadwaitingTime));
```

Comment:
@npark-mozilla I saw this assertion working in different tests, but here it does not work 100% of the times. Any idea or suggestion you may have is more than welcome, thanks!

Reply:
above assert may fail if the progressBar disappears too fast. If this is the case, I think we need a different way to check whether a new content is being displayed on webview.  For example, if we have an HTML page that prints current time using a simple JS, and access that webpage, it would display a different data upon refresh. (The webpage would need to be on a local webserver)  If so, then we can prove the page is reloaded, and won't have to worry about the spinner or progressbar.

Reply:
That's a good idea, thanks for the suggestion. Will work on that.